### PR TITLE
SoftmaxMatcher

### PR DESCRIPTION
**Pull request by xuxinyi389**
### PR Docs

无
### PR APIs

```
1.修复SoftmaxMatcher，
2.强化GenericMatcher，
3.增加获取custom运算的分支逻辑
4.对mode属性采用更严格的匹配检查
```

### CODE REVIEW

File: paconvert/transformer/basic_transformer.py [hunk 1]
```diff
@@ -225,9 +225,13 @@ def visit_Attribute(self, node):
                     ".".join(["torch.autograd.function.FunctionCtx", attr_list[-1]])
                 )
             if is_distribution_api:
-                torch_class_apis.append(
-                    ".".join(["torch.distributions.Distribution", attr_list[-1]])
-                )
+                # more strictly check on mode attribute
```
Comment (zhwesky2010) [inline paconvert/transformer/basic_transformer.py:228]: 注释直接写问题case

File: paconvert/transformer/basic_transformer.py [hunk 2]
```diff
@@ -225,9 +225,13 @@ def visit_Attribute(self, node):
...
+                if attr_list[-1] in ["mode"]:
```
Comment (zhwesky2010) [inline paconvert/transformer/basic_transformer.py:229]:
```
if full_attr != "config.mode":
    torch_class_apis.append(".".join(["paddle.Tensor", attr_list[-1]]))
```

这样就行了

不需要pass这种冗余分支

File: paconvert/api_matcher.py [hunk 3]
```diff
@@ -3827,11 +3825,51 @@ def generate_code(self, kwargs):
...
+                    ret = 1
+                return ret
+
+            original_softmax_fun = paddle.nn.functional.softmax
+
+            def softmax(x, axis = None, dtype=None, name=None):
+                if axis is None:
+                    axis = _get_softmax_dim(x.ndim)
+                return original_softmax_fun(x, axis, dtype, name)
+            setattr(paddle.nn.functional, 'softmax_custom', softmax)
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3844]: 这里能直接修改paddle.nn.functional.softmax、paddle.nn.functional.log_softmax吗

File: paconvert/api_matcher.py [hunk 4]
```diff
@@ -3827,11 +3825,51 @@ def generate_code(self, kwargs):
...
-            kwargs.pop("dim", "None")
-            kwargs["axis"] = 0
-        return GenericMatcher.generate_code(self, kwargs)
+        is_custom = False
+        if "dim" not in kwargs or kwargs["dim"] == "None":
+            self.write_aux_code()
+            kwargs["dim"] = "None"
+            if "functional" in self.get_paddle_api():
+                is_custom = True
+        return GenericMatcher.generate_code(self, kwargs, is_custom=is_custom)
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:3872]: 这个特殊的逻辑，需要改通用基类吗，最好不要改动

File: paconvert/api_matcher.py (hunk 3 above)
Comment (xuxinyi389) [reply paconvert/api_matcher.py:3844]: 这个是没办法的，单测会报错。报错内容为：循环调用超过最大深度。  保存 原始 original_softmax = paddle.nn.functional.softmax，在新注册的paddle.nn.functional.softmax中调用original_softmax会出错

File: paconvert/api_matcher.py [hunk 5]
```diff
@@ -148,6 +148,14 @@ def generate_code(self, kwargs):
         return code
 
 
+class RegistCustomMethodMatcher(GenericMatcher):
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:151]: 这样代码风格还是比较丑

对于function 的softmax，直接改成paddle.nn.functional.softmax(x, axis=paddle_aux._get_softmax_dim(x))

对于Layer 的softmax，在paddle_aux重新注册forward

**Review by zhwesky2010 — APPROVED** [review]
LGTM
